integration/container/logs_test.go

```diff
@@ -32,3 +38,145 @@ func TestLogsFollowTailEmpty(t *testing.T) {
 }
+
+func TestLogs(t *testing.T) {
+	drivers := []string{local.Name, jsonfilelog.Name}
+
+	for _, logDriver := range drivers {
+		t.Run("driver "+logDriver, func(t *testing.T) {
+			testLogs(t, logDriver)
+		})
+	}
+}
+
+func testLogs(t *testing.T, logDriver string) {
+	defer setupTest(t)()
+	client := testEnv.APIClient()
+	ctx := context.Background()
+
+	testCases := []struct {
+		desc        string
+		logOps      types.ContainerLogsOptions
+		expectedOut string
+		expectedErr string
+		tty         bool
+	}{
+		// TTY, only one output streamś
+		{
+			desc: "tty/stdout and stderr",
+			tty:  true,
+			logOps: types.ContainerLogsOptions{
+				ShowStdout: true,
+				ShowStderr: true,
+			},
+			expectedOut: "this is fineaccidents happen",
+		},
+		{
+			desc: "tty/only stdout",
+			tty:  true,
+			logOps: types.ContainerLogsOptions{
+				ShowStdout: true,
+				ShowStderr: false,
+			},
+			expectedOut: "this is fineaccidents happen",
+		},
+		{
+			desc: "tty/only stderr",
+			tty:  true,
+			logOps: types.ContainerLogsOptions{
+				ShowStdout: false,
+				ShowStderr: true,
+			},
+			expectedOut: "",
+		},
+		// Without TTY, both stdout and stderrś
+		{
+			desc: "without tty/stdout and stderr",
+			tty:  false,
+			logOps: types.ContainerLogsOptions{
+				ShowStdout: true,
+				ShowStderr: true,
+			},
+			expectedOut: "this is fine",
+			expectedErr: "accidents happen",
+		},
+		{
+			desc: "without tty/only stdout",
+			tty:  false,
+			logOps: types.ContainerLogsOptions{
+				ShowStdout: true,
+				ShowStderr: false,
+			},
+			expectedOut: "this is fine",
+			expectedErr: "",
+		},
+		{
+			desc: "without tty/only stderr",
+			tty:  false,
+			logOps: types.ContainerLogsOptions{
+				ShowStdout: false,
+				ShowStderr: true,
+			},
+			expectedOut: "",
+			expectedErr: "accidents happen",
+		},
+	}
+
+	for _, tC := range testCases {
+		tC := tC
+		t.Run(tC.desc, func(t *testing.T) {
+			t.Parallel()
+			tty := tC.tty
+			id := container.Run(ctx, t, client,
+				container.WithCmd("sh", "-c", "echo -n this is fine; echo -n accidents happen >&2"),
+				container.WithTty(tty),
+				container.WithLogDriver(logDriver),
+			)
+			defer client.ContainerRemove(ctx, id, types.ContainerRemoveOptions{Force: true})
+
+			poll.WaitOn(t, container.IsStopped(ctx, client, id), poll.WithDelay(time.Millisecond*100))
+
+			logs, err := client.ContainerLogs(ctx, id, tC.logOps)
+			assert.NilError(t, err)
+			defer logs.Close()
+
+			var stdout, stderr bytes.Buffer
+			if tty {
+				// TTY, only one output stream
+				_, err = io.Copy(&stdout, logs)
+			} else {
+				_, err = stdcopy.StdCopy(&stdout, &stderr, logs)
+			}
+			assert.NilError(t, err)
+
+			stdoutStr := stdout.String()
+			if tty && testEnv.OSType == "windows" {
+				stdoutStr = stripEscapeCodes(t, stdoutStr)
+			}
+			assert.DeepEqual(t, stdoutStr, tC.expectedOut)
+			assert.DeepEqual(t, stderr.String(), tC.expectedErr)
+		})
+	}
+}
+
+func stripEscapeCodes(t *testing.T, input string) string {
```

Comment:
I found this module [github.com/Azure/go-ansiterm](https://pkg.go.dev/github.com/Azure/go-ansiterm#section-readme) which implements the full terminal-parsing state machine. (It's already vendored, too!) If you're feeling really keen, you could use it to make the test more robust by writing a "terminal emulator" fake which writes the printed characters to a string.

Here are some tips which might help if you want to give this a try:
- "C0 commands" are the non-printable ASCII codes including backspace `\b`, CR `\r`, LF `\n`
- You can embed an interface value into a struct to make the struct values implement the interface without explicitly defining all the interface methods. Any attempts to call these methods will result in a nil-dereference panic because it's literally calling a method on the embedded nil-valued interface field. (Unless you assign a concrete value to the field, of course.) I find it a very convenient trick when writing test fakes where only a subset of the interface is required for the test.

Reply:
Never mind; turns out there's a bug in go-ansiterm's parsing of OSC commands (it incorrectly interprets a bare backslash character as the string terminator instead of `<ESC> \`) so it can't correctly strip out the Set Window Title command as it incorrectly transitions to the ground state at the backslash in `C:\`. https://go.dev/play/p/9xBvg_3_QKI

Reply:
Nice module and clever idea :) Would be good to use here if it didn't have the bug. However for now I'd go with the simple strings.Replace, since the same sequences are used most of the time. And if Windows still behaves funny, then I think we could skip the tests for Windows with TTY case and revisit this when your PR is accepted into the go-ansiterm.

Reply:
The go-ansiterm bug could feasibly be worked around by stripping the Set Window Title command sequence from the input before passing it to the parser. Since terminal commands are normally parsed character-by-character with a state machine, it must therefore be possible to write a regex which will match any possible Operating System Command sequence (which includes Set Window Title). It'd probably be less work than playing whack-a-mole with cursor movement commands.

Reply:
I don't have permission to push to your branch, so I had to push to my own fork instead. Here you go: https://github.com/corhere/moby/commit/10614e994c4fcf37d2b4d76d1f9888084d09d917